contracts/test/integration/IntegrationTestVoltV2.t.sol

```diff
@@ -0,0 +1,375 @@
+                    getDomainSeperator(),
+                    keccak256(
+                        abi.encode(
+                            volt.PERMIT_TYPEHASH(),
+                            owner,
+                            address(this),
+                            voltToPermit,
+                            0,
+                            block.timestamp
+                        )
+                    )
+                )
+            )
+        );
+
+        volt.permit(
+            owner,
+            address(this),
+            voltToPermit,
+            block.timestamp,
+            v,
+            r,
+            s
+        );
+
+        assertEq(volt.allowance(owner, address(this)), voltToPermit);
+        assertEq(volt.nonces(owner), 1);
+    }
+
+    function testPermitBadNonce() public {
+        uint256 privateKey = 0xFFF;
+        address owner = vm.addr(privateKey);
+
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+            privateKey,
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    getDomainSeperator(),
+                    keccak256(
+                        abi.encode(
+                            volt.PERMIT_TYPEHASH(),
+                            owner,
+                            address(this),
+                            1e18,
+                            1,
+                            block.timestamp
+                        )
+                    )
+                )
+            )
+        );
+
+        vm.expectRevert("Volt: unauthorized");
+        volt.permit(owner, address(this), 1e18, block.timestamp, v, r, s);
+    }
+
+    function testPermitExpired() public {
+        uint256 privateKey = 0xFFF;
+        address owner = vm.addr(privateKey);
+
+        uint256 timestamp = block.timestamp;
+
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+            privateKey,
+            keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    getDomainSeperator(),
+                    keccak256(
+                        abi.encode(
+                            volt.PERMIT_TYPEHASH(),
+                            owner,
+                            address(this),
+                            1e18,
+                            0,
+                            timestamp
+                        )
+                    )
+                )
+            )
+        );
+
+        vm.warp(timestamp + 1);
+        vm.expectRevert("Volt: signature expired");
+        volt.permit(owner, address(this), 1e18, timestamp, v, r, s);
+    }
+
+    function getDomainSeperator()
+        internal
+        view
+        returns (bytes32 domainSeparator)
+    {
+        domainSeparator = keccak256(
+            abi.encode(
+                volt.DOMAIN_TYPEHASH(),
+                keccak256(bytes(volt.name())),
+                block.chainid,
+                address(volt)
+            )
+        );
```

Comment:
would be good to see some more complex fuzz tests that take an array of addresses, and an amount, and transfer between the addresses in sequence verifying the chain of transfers, voting power and delegates transfers correctly.